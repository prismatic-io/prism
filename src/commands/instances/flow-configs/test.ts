import { Command, Flags, Args, ux } from "@oclif/core";
import { gqlRequest, gql } from "../../../graphql";

interface LogNode {
  [index: string]: unknown;
  timestamp: string;
  severity: string;
  message: string;
}

interface FetchLogsResult {
  logs: LogNode[];
  cursor: string | undefined;
  executionComplete: boolean | undefined;
}

export default class TestCommand extends Command {
  static description = "Test a Flow Config of an Instance";
  static args = {
    flowConfig: Args.string({
      description: "ID of a Flow Config to test",
      required: true,
    }),
  };

  static flags = {
    ...ux.table.flags({ only: ["extended", "columns"] }),
    tail: Flags.boolean({
      required: false,
      char: "t",
      description: "Tail logs of the flow config test run",
    }),
    payload: Flags.string({
      required: false,
      char: "p",
      description:
        "Optional JSON-formatted data payload to submit with the test",
    }),
    contentType: Flags.string({
      required: false,
      char: "c",
      description: "Optional content-type for the test payload",
    }),
  };

  async run() {
    const {
      args: { flowConfig },
      flags: { tail, payload, contentType },
    } = await this.parse(TestCommand);

    const result = await gqlRequest({
      document: gql`
        mutation testInstanceFlowConfig(
          $id: ID!
          $payload: String
          $contentType: String
        ) {
          testInstanceFlowConfig(
            input: { id: $id, payload: $payload, contentType: $contentType }
          ) {
            testInstanceFlowConfigResult {
              flowConfig {
                id
              }
              execution {
                id
              }
            }
            errors {
              field
              messages
            }
          }
        }
      `,
      variables: {
        id: flowConfig,
        payload,
        contentType,
      },
    });

    if (!tail) {
      return;
    }

    const executionId =
      result.testInstanceFlowConfig.testInstanceFlowConfigResult.execution.id;
    await this.tailLogs(executionId);
  }

  private async tailLogs(executionId: string) {
    const { flags } = await this.parse(TestCommand);

    let nextCursor: string | undefined = undefined;
    // eslint-disable-next-line no-constant-condition
    while (true) {
      await ux.wait(500);

      const result: any = await this.fetchLogs(executionId, nextCursor);
      if (result === undefined) continue;

      const { logs, cursor, executionComplete } = result;

      nextCursor = cursor;

      ux.table(
        logs,
        {
          timestamp: {},
          severity: {
            minWidth: 12,
          },
          message: {},
        },
        { ...flags, "no-header": true }
      );

      if (executionComplete) return;
    }
  }

  private async fetchLogs(
    executionId: string,
    nextCursor?: string
  ): Promise<FetchLogsResult | undefined> {
    const results = await gqlRequest({
      document: gql`
        query listInstanceTestLogs($executionId: ID!, $nextCursor: String) {
          logs(
            executionResult: $executionId
            after: $nextCursor
            orderBy: { field: TIMESTAMP, direction: ASC }
          ) {
            edges {
              node {
                timestamp
                severity
                message
              }
              cursor
            }
          }
        }
      `,
      variables: {
        executionId,
        nextCursor,
      },
    });

    const {
      edges,
    }: {
      edges: { cursor?: string; node: LogNode }[];
    } = results.logs;
    if (!edges || edges.length === 0) {
      return undefined;
    }

    const logs = edges.map(({ node }) => node);
    const executionComplete = logs.reduce<boolean>(
      (result: boolean, { message }) =>
        result || message.startsWith("Ending Instance Execution"),
      false
    );

    const { cursor } = edges[edges.length - 1];
    return { logs, cursor, executionComplete };
  }
}
