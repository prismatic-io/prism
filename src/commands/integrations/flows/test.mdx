```bash
# Test an integration flow with a payload file and tail the logs and step results
prism integrations:flows:test \
    -p=some_payload_file.xml
    -c=application/xml
    --tail-logs
    --tail-results
```
