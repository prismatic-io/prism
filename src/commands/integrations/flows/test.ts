import { Command, Flags, Args, ux } from "@oclif/core";
import { gql, gqlRequest } from "../../../graphql.js";

interface LogNode {
  [index: string]: unknown;
  timestamp: string;
  severity: string;
  message: string;
}

interface FetchLogsResult {
  logs: LogNode[];
  cursor: string | undefined;
  executionComplete: boolean;
}

export default class TestCommand extends Command {
  static description = "Run a test of an Integration Flow";
  static args = {
    flow: Args.string({ description: "ID of a flow to test", required: true }),
  };

  static flags = {
    ...ux.table.flags({ only: ["extended", "columns"] }),
    tail: Flags.boolean({
      char: "t",
      description: "Tail logs of the integration test run",
      required: false,
    }),
    payload: Flags.string({
      char: "p",
      description: "Optional JSON-formatted data payload to submit with the test",
      required: false,
    }),
    contentType: Flags.string({
      char: "c",
      description: "Optional content-type for the test payload",
      required: false,
    }),
  };

  async run() {
    const {
      args: { flow },
      flags: { tail, payload, contentType },
    } = await this.parse(TestCommand);

    const result = await gqlRequest({
      document: gql`
        mutation testIntegrationFlow(
          $id: ID!
          $payload: String
          $contentType: String
        ) {
          testIntegrationFlow(
            input: { id: $id, payload: $payload, contentType: $contentType }
          ) {
            testIntegrationFlowResult {
              flow {
                id
              }
              execution {
                id
              }
            }
            errors {
              field
              messages
            }
          }
        }
      `,
      variables: {
        id: flow,
        payload,
        contentType,
      },
    });
    const executionId = result.testIntegrationFlow.testIntegrationFlowResult.execution.id;
    console.log(`Execution ID: ${executionId}`);
    if (tail) {
      await this.tailLogs(executionId);
    }
  }

  private async tailLogs(executionId: string) {
    const { flags } = await this.parse(TestCommand);

    let nextCursor: string | undefined = undefined;
    while (true) {
      await ux.wait(500);

      const result: any = await this.fetchLogs(executionId, nextCursor);
      if (result === undefined) continue;

      const { logs, cursor, executionComplete } = result;

      nextCursor = cursor;

      ux.table(
        logs,
        {
          timestamp: {},
          severity: {
            minWidth: 12,
          },
          message: {},
        },
        { ...flags, "no-header": true },
      );

      if (executionComplete) return;
    }
  }

  private async fetchLogs(
    executionId: string,
    nextCursor?: string,
  ): Promise<FetchLogsResult | undefined> {
    const results = await gqlRequest({
      document: gql`
        query listIntegrationTestLogs($executionId: ID!, $nextCursor: String) {
          logs(
            executionResult: $executionId
            after: $nextCursor
            orderBy: { field: TIMESTAMP, direction: ASC }
          ) {
            edges {
              node {
                timestamp
                severity
                message
              }
              cursor
            }
          }
        }
      `,
      variables: {
        executionId,
        nextCursor,
      },
    });

    const { edges }: { edges: { node: LogNode; cursor?: string }[] } = results.logs;
    if (!edges || edges.length === 0) {
      return undefined;
    }

    const logs = edges.map(({ node }) => node);
    const executionComplete = logs.reduce(
      (result: boolean, { message }) => result || message.startsWith("Ending Instance"),
      false,
    );

    const { cursor } = edges[edges.length - 1];
    return { logs, cursor, executionComplete };
  }
}
