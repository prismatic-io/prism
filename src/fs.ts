import { promises as fs } from "fs";

export const exists = async (path: string) => {
  return fs.access(path).then(
    () => true,
    () => false,
  );
};

export const readStdin = async () => {
  return new Promise((resolve, reject) => {
    process.stdin
      .on("readable", () => {
        const result = process.stdin.read();
        if (result !== null) {
          process.stdin.removeAllListeners();
          resolve(result.toString());
        }
      })
      .on("error", reject);
  });
};

export { fs };
