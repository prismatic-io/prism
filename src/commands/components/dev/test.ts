import { promisify } from "util";
import dotenv from "dotenv";
import { Command, Flags, CliUx } from "@oclif/core";
import inquirer, { DistinctQuestion } from "inquirer";
import { snakeCase, upperCase, kebabCase } from "lodash";
import { serverTypes } from "@prismatic-io/spectral"; // FIXME: Get rid of this and stop exporting it in Spectral.
import {
  loadEntrypoint,
  createComponentPackage,
  publishDefinition,
  uploadConnectionIcons,
  uploadFile,
  validateDefinition,
  checkPackageSignature,
} from "../../../utils/component/publish";
import { displayLogs } from "../../../utils/execution/logs";
import {
  buildComponentTestHarnessIntegration,
  componentTestIntegrationName,
  ComponentTestInfo,
} from "../../../utils/integration/definition";
import { importDefinition } from "../../../utils/integration/import";
import {
  deleteIntegration,
  runIntegrationFlow,
} from "../../../utils/integration/invoke";
import { pollForActiveConfigVarState } from "../../../utils/integration/query";
import { Expression } from "../../../utils/integration/export";
import { exists } from "../../../fs";
import { whoAmI } from "../../../utils/user/query";
import { spawnProcess } from "../../../utils/process";
import {
  printFinalStepResults,
  writeFinalStepResults,
} from "../../../utils/execution/stepResults";
import { deleteComponentByKey } from "../../../utils/component/deleteByKey";

const setTimeoutPromise = promisify(setTimeout);

const envVarCase = (name: string): string =>
  upperCase(snakeCase(name)).replace(/\s+/g, "_");

const toInquirerInputType = (
  type: string,
  collection: string | undefined
): Required<DistinctQuestion["type"]> => {
  if (collection) {
    // FIXME: Improve prompting instead of instantly bailing to editor.
    return "editor";
  }

  switch (type) {
    case "boolean":
      return "checkbox";
    case "password":
      return "password";
    case "code":
      return "editor";
    default:
      return "input";
  }
};

const getInputQuestion = ({
  key,
  label,
  type,
  collection,
  default: defaultValue,
}: serverTypes.Input): DistinctQuestion => ({
  type: toInquirerInputType(type, collection),
  name: key,
  message: `${label}:`,
  when: (answers) => {
    const envVar = envVarCase(key);
    const exists = envVar in process.env;
    if (exists) {
      const value = process.env[envVar] ?? "";
      answers[key] = collection
        ? { type: "complex", value: JSON.parse(value) }
        : { type: "value", value };
    }
    return !exists;
  },
  filter: (value) => {
    if (type === "connection") {
      return { type: "configVar", value };
    }

    if (collection) {
      return {
        type: "complex",
        value: JSON.parse(value),
      };
    }

    return {
      type: "value",
      value,
    };
  },
  default: () => (type === "connection" ? "testConnection" : defaultValue),
});

interface PromptAnswers {
  action: serverTypes.Action;
  actionInputs: Record<string, Expression>;
  connection?: serverTypes.Connection;
  connectionInputs?: Record<string, Expression>;
}

const valuesFromAnswers = ({
  action,
  actionInputs,
  connection,
  connectionInputs = {},
}: PromptAnswers): Pick<ComponentTestInfo, "actionInfo" | "connectionInfo"> => {
  const actionValues = Object.entries(actionInputs).reduce<
    ComponentTestInfo["actionInfo"]["values"]
  >(
    (result, [key, value]) => ({
      ...result,
      [key]: value,
    }),
    {}
  );

  const connectionValues = Object.entries(connectionInputs).reduce<
    Required<ComponentTestInfo>["connectionInfo"]["values"]
  >(
    (result, [key, value]) => ({
      ...result,
      [key]: value,
    }),
    {}
  );

  const connectionInfo = connection
    ? { key: connection.key, values: connectionValues }
    : undefined;

  return {
    actionInfo: { key: action.key, values: actionValues },
    connectionInfo,
  };
};

export default class TestCommand extends Command {
  static description =
    "Run an action of a component within a test integration in the integration runner";
  static flags = {
    envPath: Flags.string({
      required: false,
      default: ".env",
      char: "e",
      description: "Path to dotenv file to load for supplying testing values",
    }),
    build: Flags.boolean({
      required: false,
      default: true,
      allowNo: true,
      char: "b",
      description: "Build the component prior to testing",
    }),
    "output-file": Flags.string({
      required: false,
      char: "o",
      description: "Output the results of the action to a specified file",
    }),
    "print-results": Flags.boolean({
      required: false,
      default: false,
      description: "Print the results of the action to stdout",
    }),
    "clean-up": Flags.boolean({
      required: false,
      default: false,
      description:
        "Clean up the integration and temporary component after running the action",
    }),
  };

  async run() {
    const {
      flags: {
        envPath,
        build,
        "output-file": outputFile,
        "print-results": printResults,
        "clean-up": cleanUp,
      },
    } = await this.parse(TestCommand);

    // Save the current working directory, so we can return later after moving to dist/
    const cwd = process.cwd();

    if (build) {
      console.log("Building component...");
      await spawnProcess(["npm", "run", "build"], {});
    }

    if (await exists(envPath)) {
      const { error } = dotenv.config({ path: envPath, override: true });
      if (error) {
        CliUx.ux.error(`Failed to load specified dotenv file: ${error}`, {
          exit: 1,
        });
      }
    }

    CliUx.ux.action.start("Validating Component");

    const { name } = await whoAmI();
    if (!name) {
      CliUx.ux.error(
        "Failed to determine the name of the currently logged in user.",
        {
          exit: 1,
        }
      );
    }

    const testingKey = kebabCase(name);

    const definition = await loadEntrypoint();
    const { key: componentKey, public: isPublic } = definition;
    definition.key = `${componentKey}-${testingKey}-testing`;
    definition.display.label = `${definition.display.label} ${name} Testing`;
    await validateDefinition(definition);

    const packagePath = await createComponentPackage();
    const signatureMatches = await checkPackageSignature(
      definition,
      packagePath
    );

    CliUx.ux.action.stop();

    if (!signatureMatches) {
      CliUx.ux.action.start("Publishing Component");

      const { iconUploadUrl, packageUploadUrl, connectionIconUploadUrls } =
        await publishDefinition(definition);

      const {
        display: { iconPath },
      } = definition;
      await uploadFile(iconPath, iconUploadUrl);
      await uploadConnectionIcons(definition, connectionIconUploadUrls);
      await uploadFile(packagePath, packageUploadUrl);

      CliUx.ux.action.stop();
    }

    const publishedTimestamp = Date.now();

    const actions = definition.actions || {};

    const { action } = await inquirer.prompt<{ action: serverTypes.Action }>({
      type: "list",
      name: "action",
      message: "Action:",
      choices: Object.entries(actions).map(
        ([
          key,
          {
            display: { label },
          },
        ]) => ({
          name: label,
          value: key,
          short: key,
        })
      ),
      default: Object.keys(actions)[0],
      filter: (value: string) => actions[value],
    });

    const { inputs } = action;

    // Ask for values of action's inputs
    const actionInputs = await inquirer.prompt(
      inputs.map((i) => getInputQuestion(i))
    );

    const answers = {
      action,
      actionInputs,
    };

    // Ask about Connection to test if there is a connection type input
    const hasConnection = inputs.some(({ type }) => type === "connection");
    if (hasConnection) {
      const connections = definition.connections || [];
      const { connection } = await inquirer.prompt<{
        connection: serverTypes.Connection;
      }>({
        type: "list",
        name: "connection",
        message: "Connection:",
        choices: connections.map(({ key, label }) => ({
          name: label,
          value: key,
          short: key,
        })),
        default: Object.keys(connections)[0],
        filter: (value) => {
          const [connection] = connections.filter(({ key }) => value === key);
          return connection;
        },
      });

      // Prompt for connection's inputs
      const { inputs } = connection;
      const connectionInputs = await inquirer.prompt(
        inputs
          .filter(({ shown }) => shown === undefined || shown === true)
          .map((i) => getInputQuestion(i))
      );

      Object.assign(answers, { connection, connectionInputs });
    }

    const { actionInfo, connectionInfo } = valuesFromAnswers(answers);

    CliUx.ux.action.start("Assembling test integration");

    // FIXME: Wait for version to be available but due to issues we have to do a static wait.
    const wait = 5000 - (Date.now() - publishedTimestamp);
    if (wait > 0) {
      await setTimeoutPromise(wait);
    }

    // Build up YAML
    const harnessYaml = await buildComponentTestHarnessIntegration({
      integrationInfo: {
        name: componentTestIntegrationName(componentKey, name),
      },
      componentInfo: { key: definition.key, isPublic: isPublic ?? false },
      actionInfo,
      connectionInfo,
    });

    CliUx.ux.action.stop();
    CliUx.ux.action.start("Updating test Integration");

    const {
      integrationId,
      flows: [{ id: flowId }],
      pendingAuthorizations,
    } = await importDefinition(harnessYaml);

    CliUx.ux.action.stop();

    // Prompt for user authorization of pending connections
    if (pendingAuthorizations.length > 0) {
      const [{ id, url }] = pendingAuthorizations;
      if (!url) {
        throw new Error(
          "Did not receive a valid URL for authorization. Verify your Connection inputs."
        );
      }

      CliUx.ux.url("Authorize URL", url);
      await CliUx.ux.anykey(
        "Press any key to open your browser and authorize the Connection"
      );
      await CliUx.ux.open(url);

      CliUx.ux.action.start("Waiting for Connection authorization");

      await pollForActiveConfigVarState(integrationId, id);

      CliUx.ux.action.stop();
    }

    CliUx.ux.action.start("Running test Integration");

    const { executionId } = await runIntegrationFlow({ integrationId, flowId });

    await displayLogs(executionId);

    if (outputFile) {
      process.chdir(cwd);
      console.log(`Writing step results to ${outputFile}`);
      await writeFinalStepResults(executionId, outputFile);
    }

    if (printResults) {
      await printFinalStepResults(executionId);
    }

    if (cleanUp) {
      CliUx.ux.action.start(`Cleaning up test Integration (${integrationId})`);
      await deleteIntegration(integrationId);
      CliUx.ux.action.stop();

      CliUx.ux.action.start(`Cleaning up test component (${definition.key})`);
      await deleteComponentByKey(definition.key);
      CliUx.ux.action.stop();
    }

    CliUx.ux.action.stop();
  }
}
