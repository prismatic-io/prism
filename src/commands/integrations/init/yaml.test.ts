import { describe, it, expect } from "bun:test";
import GenerateIntegrationFromYAMLCommand, { createInputsString } from "./yaml.js";
import fs from "fs";
import path from "path";
import { readFile } from "fs-extra";
import { walkDir } from "../../../fs";

describe("createInputsString", () => {
  const testAction = {
    action: {
      component: {
        isPublic: true,
        key: "test-component",
        version: 14,
      },
      key: "test-component-trigger",
    },
    description: "",
    inputs: {
      emptyInput: {
        type: "value",
        value: "",
      },
      stringInput: {
        type: "value",
        value: "my string",
      },
      numberInput: {
        type: "value",
        value: 1,
      },
      booleanInput: {
        type: "value",
        value: false,
      },
      keyValueListInput: {
        type: "complex",
        value: [
          {
            type: "value",
            value: "myValueA",
            name: {
              type: "value",
              value: "myKeyA",
            },
          },
          {
            type: "value",
            value: "myValueB",
            name: {
              type: "value",
              value: "myKeyB",
            },
          },
        ],
      },
      emptyKeyValueListInput: {
        type: "complex",
        value: [],
      },
      valueListInput: {
        type: "complex",
        value: ["item a", "item b", "item c"],
      },
      emptyValueListInput: {
        type: "complex",
        value: [],
      },
    },
    name: "Test action",
    isTrigger: false,
    steps: [],
  };

  describe("for triggers", () => {
    const testTrigger = {
      ...testAction,
      isTrigger: true,
    };

    const expected = `
emptyInput: { value: "", },
stringInput: { value: "my string", },
numberInput: { value: 1, },
booleanInput: { value: false, },
keyValueListInput: { value: {myKeyA: "myValueA",myKeyB: "myValueB",}, },
emptyKeyValueListInput: { value: [], },
valueListInput: { value: ["undefined","undefined","undefined",], },
emptyValueListInput: { value: [], },`;

    it("correctly converts inputs", () => {
      const resultString = createInputsString(testTrigger);
      expect(resultString).toBe(expected);
    });
  });

  describe("for actions", () => {
    const expected = `
emptyInput: "",
stringInput: "my string",
numberInput: 1,
booleanInput: false,
keyValueListInput: {myKeyA: "myValueA",myKeyB: "myValueB",},
emptyKeyValueListInput: [],
valueListInput: ["undefined","undefined","undefined",],
emptyValueListInput: [],`;

    it("correctly converts inputs", () => {
      const resultString = createInputsString(testAction);
      expect(resultString).toBe(expected);
    });
  });
});

const FLOW_GENERATION_TIMEOUT = 5 * 60 * 1000; // 5 minutes
interface SpecMeta {
  fileName: string;
  name: string;
}

describe("YAML CNI generation tests", () => {
  const basePath = process.cwd();
  const integrationsPath = path.resolve("src/commands/integrations");
  const tempPath = path.resolve(`${integrationsPath}/init/temp`);

  if (!fs.existsSync(tempPath)) {
    fs.mkdirSync(tempPath);
  }

  const specs = fs
    .readdirSync(`${integrationsPath}/init/fixtures/specs`)
    .map<SpecMeta>((fileName) => {
      const [name] = fileName.toLowerCase().split(".");
      return { fileName, name };
    });

  for (const { fileName, name } of specs) {
    describe(name, () => {
      it(
        "should generate successfully",
        async () => {
          process.chdir(tempPath);

          await GenerateIntegrationFromYAMLCommand.run([
            `--yamlFile=../fixtures/specs/${fileName}`,
          ]);

          expect(process.cwd()).toStrictEqual(tempPath);

          process.chdir(basePath);
        },
        FLOW_GENERATION_TIMEOUT,
      );

      it("should match scaffolding snapshots", async () => {
        process.chdir(tempPath);
        // TODO: Bun snapshotting has a bug where it is dropping the slash for `\.ts` in webpack.config.js.
        // Exclude package.json now that we're fetching the latest Spectral version from the registry.
        const targets = await walkDir(name, [".png", "webpack.config.js", "package.json"]);
        for (const target of targets) {
          const contents = await readFile(target, "utf-8");
          expect(contents).toMatchSnapshot(target);
        }

        process.chdir(basePath);
      });
    });
  }
});
