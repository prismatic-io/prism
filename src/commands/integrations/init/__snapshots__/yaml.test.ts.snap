// Bun Snapshot v1, https://goo.gl/fbAQLP

exports[`YAML CNI generation tests simple-integration should match scaffolding snapshots: simple-integration/jest.config.js 1`] = `
"module.exports = {
  preset: "ts-jest",
  testEnvironment: "node",
};
"
`;

exports[`YAML CNI generation tests simple-integration should match scaffolding snapshots: simple-integration/.npmrc 1`] = `
"@component-manifests:registry=https://app.dev.prismatic-dev.io/packages/npm
"
`;

exports[`YAML CNI generation tests simple-integration should match scaffolding snapshots: simple-integration/.prettierignore 1`] = `
"# Ignore artifacts:
build
coverage
"
`;

exports[`YAML CNI generation tests simple-integration should match scaffolding snapshots: simple-integration/.prettierrc 1`] = `
"{}
"
`;

exports[`YAML CNI generation tests simple-integration should match scaffolding snapshots: simple-integration/tsconfig.json 1`] = `
"{
  "compilerOptions": {
    "target": "es6",
    "lib": ["esnext"],
    "strict": true,
    "esModuleInterop": true,
    "forceConsistentCasingInFileNames": true,
    "module": "esnext",
    "moduleResolution": "node"
  },
  "include": ["src", ".spectral/*"]
}
"
`;

exports[`YAML CNI generation tests simple-integration should match scaffolding snapshots: simple-integration/.spectral/index.ts 1`] = `
"import type { ComponentManifest, ConfigPage, ScopedConfigVar } from "@prismatic-io/spectral";

// @ts-ignore
import type { configPages, componentRegistry, userLevelConfigPages, scopedConfigVars } from "../src";

type IsAny<T> = 0 extends 1 & T ? true : false;

type TConfigPages =
  IsAny<typeof configPages> extends true
    ? { [key: string]: ConfigPage }
    : typeof configPages;

type TUserLevelConfigPages =
  IsAny<typeof userLevelConfigPages> extends true
    ? { [key: string]: ConfigPage }
    : typeof userLevelConfigPages;

type TComponentRegistry =
  IsAny<typeof componentRegistry> extends true
    ? { [key: string]: ComponentManifest }
    : typeof componentRegistry;

type TScopedConfigVarMap =
  IsAny<typeof scopedConfigVars> extends true
    ? { [key: string]: ScopedConfigVar }
    : typeof scopedConfigVars;

declare module "@prismatic-io/spectral" {
  // eslint-disable-next-line @typescript-eslint/no-empty-interface
  interface IntegrationDefinitionConfigPages extends TConfigPages {}

  // eslint-disable-next-line @typescript-eslint/no-empty-interface
  interface IntegrationDefinitionUserLevelConfigPages
    extends TUserLevelConfigPages {}

  // eslint-disable-next-line @typescript-eslint/no-empty-interface
  interface IntegrationDefinitionComponentRegistry extends TComponentRegistry {}

  // eslint-disable-next-line @typescript-eslint/no-empty-interface
  interface IntegrationDefinitionScopedConfigVars extends TScopedConfigVarMap {}
}

"
`;

exports[`YAML CNI generation tests simple-integration should match scaffolding snapshots: simple-integration/flows/basic.ts 1`] = `
"import { flow } from "@prismatic-io/spectral";

export const basic = flow({
  name: "Basic",
  stableKey: "basic",
  description: "",
  isSynchronous: false,
  endpointSecurityType: "customer_optional",
  onTrigger: {
    component: "webhook-triggers",
    key: "webhook",
    values: {
      
body: { value: false, },
contentType: { value: 7, },
headers: { value: [], },
testObj: { value: {someKey: "someValue",someKey2: "someValue2",}, },
    },
  },
  onExecution: async (context, params) => {
    context.components.log.writeLog({
      
level: "info",
message: "Log message 1",
    });
    context.components.helloWorldComponent.helloAction({
      
arrayTest: ["listItem1","listItem2","listItem3",],
kvTest: {someKey: "someValue",someKey2: "someValue2",},
todoListId: "",
    });

    // @TODO - what's a reasonable return object?
    return {
      data: [],
    };
  },
});

export default basic;
"
`;

exports[`YAML CNI generation tests simple-integration should match scaffolding snapshots: simple-integration/src/client.ts 1`] = `
"import type { Connection } from "@prismatic-io/spectral";

export const createClient = (connection: Connection) => {
  // Create a client using the provided Connection for the
  // service you're consuming from this Integration.
  return {
    call: async (name: unknown) =>
      Promise.resolve(\`Hello, \${name} using connection \${connection.key}\`),
  };
};
"
`;

exports[`YAML CNI generation tests simple-integration should match scaffolding snapshots: simple-integration/src/componentRegistry.ts 1`] = `
"import { componentManifests } from "@prismatic-io/spectral";

export const componentRegistry = componentManifests({
  // this is where you'll register your component manifests
});
"
`;

exports[`YAML CNI generation tests simple-integration should match scaffolding snapshots: simple-integration/src/index.ts 1`] = `
"import { integration } from "@prismatic-io/spectral";
import flows from "./flows";
import { configPages } from "./configPages";
import { componentRegistry } from "./componentRegistry";

export { configPages } from "./configPages";
export { componentRegistry } from "./componentRegistry";

export default integration({
  name: "Simple Integration",
  description: "",
  iconPath: "icon.png",
  flows,
  configPages,
  componentRegistry,
});
"
`;

exports[`YAML CNI generation tests simple-integration should match scaffolding snapshots: simple-integration/src/flows/index.ts 1`] = `
"TODO
"
`;
